service/matching/ratelimit_manager.go

```diff
@@ -0,0 +1,188 @@
+		r.rateLimitSource = rateLimitSource
+	} else {
+		r.effectiveRPS = r.systemRPS
+		r.rateLimitSource = enumspb.RATE_LIMIT_SOURCE_SYSTEM
+	}
+}
+
+// Lazy injection of poll metadata.
+// Internally call UpdateRateLimit to share the same mutex.
+func (r *rateLimitManager) InjectWorkerRPS(meta *pollMetadata) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	var rps *float64
+	if meta != nil && meta.taskQueueMetadata != nil {
+		if workerRPS := meta.taskQueueMetadata.GetMaxTasksPerSecond(); workerRPS != nil {
+			value := workerRPS.GetValue()
+			rps = &value
+		}
+	}
+	r.workerRPS = rps
+	// updateRatelimitLocked includes internal logic to determine if an update is needed,
+	// so calling it unconditionally is safe and avoids redundant updates.
+	r.updateRatelimitLocked()
+}
+
+// Return the effective RPS and its source together.
+// The source can be API, worker or system default.
+// It defaults to the system dynamic config.
+func (r *rateLimitManager) GetEffectiveRPSAndSource() (float64, enumspb.RateLimitSource) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.computeEffectiveRPSAndSourceLocked()
+	return r.effectiveRPS, r.rateLimitSource
+}
+
+func (r *rateLimitManager) GetRateLimiter() quotas.RateLimiter {
+	return r.dynamicRateLimiter
+}
+
+// Updates the API-configured RPS based on the latest user data
+// and applies the new rate limit if the effective RPS has changed.
+func (r *rateLimitManager) UserDataChanged() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// Immediately recompute and apply rate limit if necessary.
+	r.updateRatelimitLocked()
+}
+
+// trySetRPSFromUserDataLocked sets the apiConfigRPS from user data.
+func (r *rateLimitManager) trySetRPSFromUserDataLocked() {
+	userData, _, err := r.userDataManager.GetUserData()
+	if err != nil {
+		return
+	}
+	config := userData.GetData().GetPerType()[int32(r.taskQueueType)].GetConfig()
+	if config == nil {
+		return
+	}
+	// If rate limit is an empty message, it means rate limit could have been unset via API.
+	// In this case, the apiConfigRPS will need to be unset.
+	rateLimit := config.GetQueueRateLimit()
+	if rateLimit.GetRateLimit() == nil {
+		r.apiConfigRPS = nil
+		return
+	}
+	val := float64(rateLimit.GetRateLimit().GetRequestsPerSecond())
+	r.apiConfigRPS = &val
+}
+
+// updateRatelimitLocked checks and updates the rate limit if changed.
+func (r *rateLimitManager) updateRatelimitLocked() {
+	// Always check for api configured rate limits before any update to the rate limits.
+	r.trySetRPSFromUserDataLocked()
+	oldRPS := r.effectiveRPS
+	r.computeEffectiveRPSAndSourceLocked()
+	newRPS := r.effectiveRPS
+	if oldRPS == newRPS {
+		// No update required
+		return
+	}
+	effectiveRPSPartitionWise := r.effectiveRPS / r.getNumberOfReadPartitions()
+	burst := int(math.Ceil(effectiveRPSPartitionWise))
+	burst = max(burst, r.config.MinTaskThrottlingBurstSize())
+	r.dynamicRateBurst.SetRPS(effectiveRPSPartitionWise)
+	r.dynamicRateBurst.SetBurst(burst)
+	r.forceRefreshRateOnce.Do(func() {
```

Comment:
I thought there was a subtle problem here where we would spend our "Once" on the first update from either config or from polls, and then fail to respond to the first update from the other one. I think it turns out that's not really a problem, since the config should come before any polls, and we want the config to take precedence.

anyway, I'm thinking we should get rid of the Once and just always Refresh here: we already checked if the value changed, so we only get here if we have a new value, and then why wouldn't we want it to be applied immediately? the only situation might be if workers disagree and are sending values such that it flips back and forth.

I guess we can leave it for now until we consolidate on just using simpleLimiters?